src/Microsoft.VisualStudio.Editor.Razor/RazorDirectiveCompletionSource.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Core.Imaging;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    internal class RazorDirectiveCompletionSource : IAsyncCompletionSource
+    {
+        // Internal for testing
+        internal static readonly object DescriptionKey = new object();
+        internal static readonly AccessibleImageId DirectiveImageGlyph = new AccessibleImageId(
+            new ImageId(KnownImageIds.ImageCatalogGuid, KnownImageIds.Type),
+            "Razor Directive.");
+        internal static readonly ImmutableArray<CompletionFilter> DirectiveCompletionFilters = new[] {
+            new CompletionFilter("Razor Directive", "r", DirectiveImageGlyph)
+        }.ToImmutableArray();
+        private static readonly IEnumerable<DirectiveDescriptor> DefaultDirectives = new[]
+        {
+            CSharpCodeParser.AddTagHelperDirectiveDescriptor,
+            CSharpCodeParser.RemoveTagHelperDirectiveDescriptor,
+            CSharpCodeParser.TagHelperPrefixDirectiveDescriptor,
+        };
+
+        // Internal for testing
+        internal readonly VisualStudioRazorParser _parser;
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+
+        public RazorDirectiveCompletionSource(
+            VisualStudioRazorParser parser,
+            ForegroundDispatcher foregroundDispatcher)
+        {
+            if (parser == null)
+            {
+                throw new ArgumentNullException(nameof(parser));
+            }
+
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            _parser = parser;
+            _foregroundDispatcher = foregroundDispatcher;
+        }
+
+        public Task<CompletionContext> GetCompletionContextAsync(
+            CompletionTrigger trigger, 
+            SnapshotPoint triggerLocation, 
+            SnapshotSpan applicableSpan, 
+            CancellationToken token)
+        {
+            _foregroundDispatcher.AssertBackgroundThread();
+
+            var syntaxTree = _parser.CodeDocument?.GetSyntaxTree();
+            if (!AtDirectiveCompletionPoint(syntaxTree, triggerLocation))
+            {
+                return Task.FromResult(new CompletionContext(ImmutableArray<CompletionItem>.Empty));
+            }
+
+            var completionItems = GetCompletionItems(syntaxTree);
```

Comment:
We take the trigger location because we want to operate on where the user wanted completion information from. In the case of no leading `@` we do not want to provide completions because the user is outside of a valid Razor directive context.

Reply:
I see - after we skip the `if (!AtDirectiveCompletionPoint(syntaxTree, triggerLocation))` branch, it is guaranteed that we are in a correct spot to return the directives. The directives don't depend on the location in code, so we don't need to pass neither `triggerLocation` nor `applicableSpan` to `GetCompletionItems`. Is this correct?

Reply:
Correct.